### PR TITLE
[BEAM-3030] Adds a deduplication key to Watch, and uses it to handle growing files in FileIO.match

### PR DESCRIPTION
**Pull request by jkff**
https://issues.apache.org/jira/browse/BEAM-3030

CC: @reuvenlax

### CODE REVIEW

**Comment by jkff** [comment]
Fixed test failure, PTAL.

**Comment by jkff** [comment]
Test failure seems like an infrastructure flake.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Watch.java [hunk 1]
```diff
@@ -910,18 +982,19 @@ synchronized int addNewAsPending(Growth.PollResult<OutputT> pollResult) {
           "Should have drained all old pending outputs before adding new, "
               + "but there are %s old pending outputs",
           state.pending.size());
-      List<TimestampedValue<OutputT>> newPending = Lists.newArrayList();
+      Map<HashCode, TimestampedValue<OutputT>> newPending = Maps.newHashMap();
```
Comment (chamikaramj) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Watch.java]: Please add a comment about "newPending".

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Watch.java [hunk 2]
```diff
@@ -559,24 +583,34 @@ public String toString(KV<FirstStateT, SecondStateT> state) {
...
+
+      abstract Builder<InputT, OutputT, KeyT> setOutputCoder(Coder<OutputT> outputCoder);
 
-      abstract Builder<InputT, OutputT> setOutputCoder(Coder<OutputT> outputCoder);
+      abstract Growth<InputT, OutputT, KeyT> build();
+    }
 
-      abstract Growth<InputT, OutputT> build();
+    /** Specifies the coder for the output key. */
+    public Growth<InputT, OutputT, KeyT> withKeyCoder(Coder<KeyT> keyCoder) {
```
Comment (chamikaramj) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Watch.java]: s/withKeyCoder/withOutputKeyCoder and s/keyCoder/outputKeyCoder ?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Watch.java [hunk 3]
```diff
@@ -803,14 +870,18 @@ public String toString(Growth.TerminationCondition<?, TerminationStateT> termina
...
+        GrowthState<OutputT, KeyT, TerminationStateT> state,
+        Growth.TerminationCondition<?, TerminationStateT> terminationCondition) {
       this.coderFunnel =
           new Funnel<OutputT>() {
             @Override
             public void funnel(OutputT from, PrimitiveSink into) {
               try {
-                outputCoder.encode(from, Funnels.asOutputStream(into));
+                keyCoder.encode(
+                    keyFn.apply(from), Funnels.asOutputStream(into));
```
Comment (chamikaramj) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Watch.java]: Add a comment here ? (this seems to be the crux of the implementation).

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Watch.java (hunk 2 above)
Comment (jkff) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Watch.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Watch.java (hunk 3 above)
Comment (jkff) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Watch.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Watch.java (hunk 1 above)
Comment (jkff) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Watch.java]: Done.

**Pull request closed without merging**